vere/dawn.c

```diff
@@ -0,0 +1,444 @@
+      u3_lo_bail();
+    }
+
+    url = u3_nul;
+  }
+  else {
+    //  XX call de-purl directly
+    //
+    u3_noun par = u3v_wish("auru:de-purl:html");
+    u3_noun lur = u3i_string(u3_Host.ops_u.eth_c);
+    u3_noun rul = u3dc("rush", u3k(lur), u3k(par));
+
+    if ( u3_nul == rul ) {
+      if ( c3__czar == rac ) {
+        fprintf(stderr, "boot: galaxy requires ethereum gateway via -e\r\n");
+        u3_lo_bail();
+      }
+
+      url = u3_nul;
+    }
+    else {
+      //  XX revise for de-purl
+      //  auru:de-purl:html parses to (pair user purl)
+      //  we need (unit purl)
+      //
+      url = u3nc(u3_nul, u3k(u3t(u3t(rul))));
+    }
+
+    u3z(par); u3z(lur); u3z(rul);
+  }
+
+  return url;
+}
+
+/* _dawn_turf(): override contract domains with -H
+*/
+static u3_noun
+_dawn_turf(c3_c* dns_c)
+{
+  u3_noun tuf;
+
+  u3_noun par = u3v_wish("thos:de-purl:html");
+  u3_noun dns = u3i_string(dns_c);
+  u3_noun rul = u3dc("rush", u3k(dns), u3k(par));
+
+  if ( (u3_nul == rul) || (c3n == u3h(u3t(rul))) ) {
+    fprintf(stderr, "boot: invalid domain specified with -H %s\r\n", dns_c);
+    // bails, won't return
+    u3_lo_bail();
+    return u3_none;
+  }
+  else {
+    fprintf(stderr, "boot: overriding network domains with %s\r\n", dns_c);
+    u3_noun dom = u3t(u3t(rul));
+    tuf = u3nc(u3k(dom), u3_nul);
+  }
+
+  u3z(par); u3z(dns); u3z(rul);
+
+  return tuf;
+}
+
+/* u3_dawn_vent(): validatated boot event
+*/
+u3_noun
+u3_dawn_vent(u3_noun seed)
+{
+  u3_noun url, bok, pon, zar, tuf, sap;
+
+  u3_noun ship = u3h(seed);
+  u3_noun rank = u3do("clan:title", u3k(ship));
+
+  //  load snapshot if exists
+  //
+  if ( 0 != u3_Host.ops_u.ets_c ) {
+    fprintf(stderr, "boot: loading ethereum snapshot\r\n");
+    u3_noun raw_snap = u3ke_cue(u3m_file(u3_Host.ops_u.ets_c));
+    sap = u3nc(u3_nul, raw_snap);
+  }
+  else {
+    sap = u3_nul;
+  }
+
+  url = _dawn_purl(rank);
+
+  //  XX require https?
+  //
+  c3_c* url_c = ( 0 != u3_Host.ops_u.eth_c ) ?
+    u3_Host.ops_u.eth_c :
+    "https://ropsten.infura.io/v3/196a7f37c7d54211b4a07904ec73ad87";
+
+  //  pin block number
+  //
+  if ( c3y == u3_Host.ops_u.etn ) {
+    fprintf(stderr, "boot: extracting block from snapshot\r\n");
+    bok = u3do("bloq:snap:dawn", u3k(u3t(sap)));
+  }
+  else {
+    fprintf(stderr, "boot: retrieving latest block\r\n");
+
+    u3_noun oct = u3v_wish("bloq:give:dawn");
+    u3_noun kob = _dawn_eth_rpc(url_c, u3k(oct));
+    bok = u3do("bloq:take:dawn", u3k(kob));
+
+    u3z(oct); u3z(kob);
+  }
+
+  {
+    //  +hull:constitution:ethe: on-chain state
+    //
+    u3_noun hul;
+
+    if ( c3y == u3_Host.ops_u.etn ) {
+      hul = u3dc("hull:snap:dawn", u3k(ship), u3k(u3t(sap)));
+    }
+    else if ( c3__pawn == rank ) {
+      //  irrelevant, just bunt +hull
+      //
+      hul = u3v_wish("*hull:constitution:ethe");
+    }
+    else {
+      u3_noun oct;
+
+      if ( c3__earl == rank ) {
+        u3_noun seg = u3do("^sein:title", u3k(ship));
```

Comment:
Great. `^` syntax is migrating into the interpreter.

Reply:
Yeah, I think that's an intentional naming convention.  `sein:title` is "ask jael for their sponsor" and `^sein:title` is "guess based on their prefix".  Nobody should use `^sein:title` except possibly during initial boot.  Honestly, even here we should check what is our sponsor on chain (or in the snapshot) rather than just assuming it to be our biological parent.  Otherwise if your parent cuts you off, you can't boot in a new continuity era.

Reply:
This case is specifically handling moons -- we need to know the prefix parent to retrieve the appropriate keys and validate the deed signature. We could just do the equivalent of `(end 5 1 ship)` I guess.